### ISSUE TITLE
Stats and Queue Enhancement

### ISSUE THREAD

**Issue by marshalleq** [HaveAGitGat/Tdarr#168]
**Problem Summary:**
It's currently quite challenging to understand the status of the library.  This Enhancement suggestion is a first draft at attempting to clean this up.  This is driven by me having a large library with many months of encoding remaining and finding it nearly impossible to understand the status of the encoder as to where it is in the queue and how long it has left to complete.

The issue with 'freezing' stats that myself and a few others have had hasn't helped, but these changes should at least enable us to reset our libraries and confirm the progress of the encoder against what is remaining in the library.  

**Describe the solution you'd like**

Example Scenario:  Converting an entire library to h265 video.

**What is useful to know?**
1 How much of my library is in h265 already?
2 How many more non-h265 files are there to complete?
3 Where is my encoding position in the queue against these two metrics?
4 What files have I completed in this encoding run, vs previous.

**Problem Details:**
The queue includes the h265 files that are already encoded / pre-existing.  The queue only marks them as not-required when the encoder has reached that point in the  and then it may skip forward by hundreds of files, suddenly changing all the stats.  Only at this point they are skipped and added to the not required section in the chart.  So I can’t easily tell how many more encodes there are to complete.  And this is only when the stats are working, not frozen like seems to have happened / (or is happening not sure).

**To resolve this, I would propose along the following lines:**

**Queue Logic**
1 A change is made so that the Transcode queue on the Tdarr page excludes all upcoming files in the target codec if the ‘Don’t transcode videos in this codecs’ checkbox is selected.

**Charts**
On the stats pages the following adjustments are made:

1. The tabs be renamed to Queue, All, Library 1, Library 2, Lifetime
2. The Tdarr Status section under the current All, Library 1 & Library 2 is moved and renamed to Tdarr Queue Status for clarity and moved to its own tab called ‘Queue’, thus each library stat no longer has any dynamic stats of the active queue, other than updating the existing Video and Audio files stats contained within the library.
3. The Video and Audio Stats sections on the stats page are renamed to Library Stats - Video and Library Stats - Audio, (or some other means of indicating that these are on disk library stats only - e.g. A library tab heading with sub libraries underneath on a single page).
4. That the new Tdarr Queue Status charts are specific to the active queue only and do not retain any history e.g. when Tdarr is restarted.  The lost functionality here moving to the lifetime tab
5. That the all tab is renamed to ‘lifetime’ or ‘life’ to show the lifetime encodes of the installation / database
6. That a new tab is created for Library Summary to replace the All tab because All is ambiguous when you consider there’s a queue and a library to make stats on.
7. The health chart currently under Tdarr Status is moved to the new All Library section or summaries under each library as appropriate to the new design.  This should not interfere with the health check that is run when new files are added.
8. Also that the scan of new files pending for the queue ocurrs before encoding starts (if it doesn't already), or at least runs in the background and other stats update in real time accordingly.

That's my idea anyway, hopefully you like it. :D

**Comment by marshalleq** [comment]
A further enhancement to this would be that when running a schedule, the queue still shows when paused (i.e. the schedule dictates to stop processing).  This way the queue will continue to be accurate, and you can continue to change it.  Also, having a 'paused due to scheduling rules'  red flag somewhere on the queue page would be beneficial.

**Comment by zacharyd3** [comment]
As an added note to this regard. I had a test library when I first set it up to make sure everything was running as expected. I deleted this library but the stats never left my "all" page. I've also cleared a few more libraries (not deleted them, just hit reset stats) but their stats still show under "all" also. We need a way to reset the "all" tally too. My individual libraries show as 0 files processed and 0GB saved, but my all tab shows 2500+ files processed and 43GB saved still. It turns out I had a plugin error so it was stuck in a loop and "processing" the same 2 files so the stats were way off, hence clearing them out.

**Comment by HaveAGitGat** [comment]
@zacharyd3 added that a separate issue:
https://github.com/HaveAGitGat/Tdarr/issues/178

**Comment by HaveAGitGat** [comment]
> The queue includes the h265 files that are already encoded / pre-existing. The queue only marks them as not-required when the encoder has reached that point in the and then it may skip forward by hundreds of files, suddenly changing all the stats. Only at this point they are skipped and added to the not required section in the chart. So I can’t easily tell how many more encodes there are to complete. And this is only when the stats are working, not frozen like seems to have happened / (or is happening not sure).

@marshalleq I covered this a few months ago. Plugins can be made to do anything and everything. With this in mind, it is not possible to determine exactly how much work is left for Tdarr to do as the tasks plugins cover are so varied, with some files needing multiple transcode/remux passes etc. If Tdarr was built just for the single purpose of transcoding files into h265 then what you ask would be straight forward, but Tdarr isn't built for that. It allows using imported handbrake and ffmpeg presets, audio/video/subtitle transcoding/remuxing, complex custom handbrake/ffmpeg commands, moving/reading files, interacting with 3rd party APIs etc. 

If someone is using a default HandBrake preset, but then uses a plugin to add subtitles in a certain format, but that plugin needs to change the container due to the subtitle codec being unsupported in the container chosen in the HandBrake plugin, and then a similar set of events happen for audio but the audio plugin says to skip if the audio is already in aac or ac3 or dts but not if there's no stereo aac etc. You can see pre-determining the workload remaining quickly becomes near impossible.

Regarding your comments on the charts, this confused me when you mentioned it a few months ago and it still confuses me as it seems you're struggling to understand the charts. If you're just transcoding into h265, the charts on the 'All' stats page already show the workload remaining.

For example:
![image](https://user-images.githubusercontent.com/43864057/76642880-f07e8f00-654b-11ea-9692-fd493ea4f3e0.png)

In that chart, which represents the whole library's codecs, there are 4276 files in h264 and 1546 already in hevc. So Tdarr has 4276 files left to process into hevc. It's that simple. 

The stats page charts are regenerated from the entire library every 30 seconds or so, they do not rely on any history. So if Tdarr is working correctly then the stats page charts accurately reflect the current standing of the library. Use the stats tab to see how much real work is left, not the number of items in the queue which Tdarr is yet to process.

I've read through your subsequent comments but, because you don't seem to understand what the pie charts show, several points don't make sense and seem to be based on ideas you've made up on what the stats pages show, which lead to suggestions which overcomplicate some simple pie charts.

This issue seems to be less about suggestions and more about understanding what is already there.

**Comment by marshalleq** [comment]
HI thanks for reading, I'm overseas right now, I'll try go back and re-read what I wrote and see if I can relay it differently, or concede any points upon which I am in error.  Fundamentally though, I do think these charts could use improvement to be more obvious / accurate as to what they are telling us.  But who knows, perhaps I'm just being an idiot. :D

**Comment by HaveAGitGat** [comment]
Ok no problem. The stats make no sense to you due to the stats issue you've been having so will need to sort that out first.

**Comment by marshalleq** [comment]
Sorry, but that is purely an assumption on your part.  We don't know if the stats are still a problem for me or not and there's not really an easy way of finding out.  What the previous stats issue has done is lost my confidence to trust them or not.  

I am still stuck overseas, returning Sunday hopefully.  Re-reading my comments I do still strongly support them, but it's probably easier to go over that on discord.  I realise there is a lot of differing enhancements above, I actually took several days to write them so they were fairly well thought through too and based on many months of using the product.

Anyway, I'll leave this til I get back - all I'm aiming for is an easy to understand stats section that makes good sense to everyone and is laid out to match.  It's one area of Tdarr that has not as far as I know had any love since the very early days, so it wouldn't be surprising if we could find an enhancement or two. :)

**Comment by HaveAGitGat** [comment]
@marshalleq do you see any errors like this one:
https://github.com/HaveAGitGat/Tdarr/issues/162

Should be fixed in 1.107 when it's released but you'll have to re-scan your files to remove the massive error logs.

**Comment by marshalleq** [comment]
Hey, we got stuck in Australia, so I'm away from all my stuff and I have other pressing priorities.  NZ government decided to redefine the legal meaning of immediate family and exclude mothers and turned her away at the airport.  So we had to stay behind to support her (69 year old Colombian that doesn't speak any english, and Colombian borders closed - not really a choice is it).  Crazy times.  I do have remote access.  I think what I'll do is shift to the AIO image for testing purposes.  But I think also it's important to point out that this ticket was meant to be a visual improvement rather than a bug fix.  It was a lot to take in I know and I'm sure there would be debate around the specifics of it, but I figure if I can get some agreement around the intent and the logic of how things are displayed, it would benefit everyone.  But as I said, Priority 1 now is food / roof / job / other stuff.
